openquake/risklib/reinsurance.py

```diff
@@ -41,14 +42,72 @@
 DEBUG = False
 
 
-def check_fields(fields, dframe, idxdict, fname):
+def check_fields(fields, dframe, idxdict, fname, policyfname, treaties,
+                 treaty_types):
     """
     :param fields: fields to check (the first field is the primary key)
     :param dframe: DataFrame with the contents of fname
     :param idxdict: dictionary key -> index (starting from 1)
-    :param fname: file containing the fields to check
+    :param fname: file xml containing the fields to check
+    :param policyfname: file csv containing the fields to check
+    :param treaties: treaty names
+    :param treaty_types: treaty types
     """
     key = fields[0]
+    if len(dframe[key].unique()) < len(dframe):
+        raise InvalidFile(f'{policyfname}: {key} contains duplicates')
+    for treaty in treaties:
+        if treaty not in dframe.columns:
+            raise InvalidFile(f'{policyfname}: {treaty} is missing')
+    policies_from_exposure = list(idxdict)[1:]  # discard '?'
+    policies_from_csv = list(dframe.policy)
+    for policy in policies_from_exposure:
+        if policy not in policies_from_csv:
+            raise InvalidFile(f'{policyfname}: policy "{policy}" is missing')
+    if not is_numeric_dtype(dframe.liability):
+        raise InvalidFile(f'{policyfname}: liabilities must be numeric')
+    if not is_numeric_dtype(dframe.deductible):
+        raise InvalidFile(f'{policyfname}: deductibles must be numeric')
+    [indices] = np.where(dframe.liability.to_numpy() < 0)
+    if len(indices) > 0:
+        # NOTE: reporting only the first row found
+        raise InvalidFile(
+            '%s (row %d): a negative liability was found' % (
+                policyfname, indices[0] + 2))
+    [indices] = np.where(dframe.deductible.to_numpy() < 0)
+    if len(indices) > 0:
+        # NOTE: reporting only the first row found
+        raise InvalidFile(
+            '%s (row %d): a negative deductible was found' % (
+                policyfname, indices[0] + 2))
+    prop_treaties = []
+    for treaty, treaty_type in zip(treaties, treaty_types):
+        if treaty_type == 'prop':
+            prop_treaties.append(treaty)
+        else:
+            [indices] = np.where(~dframe[treaty].isin([0, 1]).to_numpy())
+            if len(indices) > 0:
+                # NOTE: reporting only the first row found
+                raise InvalidFile(
+                    '%s (row %d): values for %s must be either 0 or 1' % (
+                        policyfname, indices[0] + 2, treaty))
+    sums = np.zeros(len(dframe))
+    for prop_treaty in prop_treaties:
+        fractions = dframe[prop_treaty].to_numpy()
+        [indices] = np.where(fractions < 0)
```

Comment:
Also fractions > 1 is an error, please add that check.

Reply:
Done